src/connector/src/sink/snowflake.rs

```diff
@@ -0,0 +1,279 @@
+    /// reference: <https://docs.snowflake.com/en/developer-guide/sql-api/authenticating>
+    #[serde(rename = "snowflake.rsa_public_key_fp")]
+    pub rsa_public_key_fp: String,
+
+    /// The rsa pem key *without* encryption
+    #[serde(rename = "snowflake.private_key")]
+    pub private_key: String,
+
+    /// The s3 bucket where intermediate sink files will be stored
+    #[serde(rename = "snowflake.s3_bucket")]
+    pub s3_bucket: String,
+
+    /// s3 credentials
+    #[serde(rename = "snowflake.aws_access_key_id")]
+    pub aws_access_key_id: String,
+
+    /// s3 credentials
+    #[serde(rename = "snowflake.aws_secret_access_key")]
+    pub aws_secret_access_key: String,
+
+    /// The s3 region, e.g., us-east-2
+    #[serde(rename = "snowflake.aws_region")]
+    pub aws_region: String,
+}
+
+#[serde_as]
+#[derive(Clone, Debug, Deserialize, WithOptions)]
+pub struct SnowflakeConfig {
+    #[serde(flatten)]
+    pub common: SnowflakeCommon,
+}
+
+impl SnowflakeConfig {
+    pub fn from_hashmap(properties: HashMap<String, String>) -> Result<Self> {
+        let config =
+            serde_json::from_value::<SnowflakeConfig>(serde_json::to_value(properties).unwrap())
+                .map_err(|e| SinkError::Config(anyhow!(e)))?;
+        Ok(config)
+    }
+}
+
+#[derive(Debug)]
+pub struct SnowflakeSink {
+    pub config: SnowflakeConfig,
+    schema: Schema,
+    pk_indices: Vec<usize>,
+    is_append_only: bool,
+}
+
+impl Sink for SnowflakeSink {
+    type Coordinator = DummySinkCommitCoordinator;
+    type LogSinker = LogSinkerOf<SnowflakeSinkWriter>;
+
+    const SINK_NAME: &'static str = SNOWFLAKE_SINK;
+
+    async fn new_log_sinker(&self, writer_param: SinkWriterParam) -> Result<Self::LogSinker> {
+        Ok(SnowflakeSinkWriter::new(
+            self.config.clone(),
+            self.schema.clone(),
+            self.pk_indices.clone(),
+            self.is_append_only,
+        )
+        .await
+        .into_log_sinker(writer_param.sink_metrics))
+    }
+
+    async fn validate(&self) -> Result<()> {
+        Ok(())
+    }
+}
+
+impl TryFrom<SinkParam> for SnowflakeSink {
+    type Error = SinkError;
+
+    fn try_from(param: SinkParam) -> std::result::Result<Self, Self::Error> {
+        let schema = param.schema();
+        let config = SnowflakeConfig::from_hashmap(param.properties)?;
+        Ok(SnowflakeSink {
+            config,
+            schema,
+            pk_indices: param.downstream_pk,
+            is_append_only: param.sink_type.is_append_only(),
+        })
+    }
+}
+
+pub struct SnowflakeSinkWriter {
+    config: SnowflakeConfig,
+    schema: Schema,
+    pk_indices: Vec<usize>,
+    is_append_only: bool,
+    /// the client used to send `insertFiles` post request
+    http_client: SnowflakeHttpClient,
+    /// the client to insert file to external storage (i.e., s3)
+    s3_client: SnowflakeS3Client,
+    row_encoder: JsonEncoder,
+    row_counter: u32,
+    payload: String,
```

Comment:
Will `byte array` be encoded into a legal UTF-8 string?

Reply:
If the `byte array` contains binary data that isn't inherently UTF-8 encoded text, then I'm assuming it could not be encoded.

Since `Starrocks Sink` is using the same encoder for the payload, I think it should be good in most case, cc @xxhZs.

Reply:
Regarding starrock as well as other sinks, the possibility of conversion will be investigated afterwards, So snow, if it supports byte -> string, can implement it directly

Reply:
update: `JsonEncoder` will automatically convert any byte array to `base64` if ever encounter, as below.
https://github.com/risingwavelabs/risingwave/blob/dfe5f310417909ca4612d468577e541724d4804f/src/connector/src/sink/encoder/json.rs#L305